l10n_nl_tax_statement/models/l10n_nl_vat_statement.py

```diff
@@ -0,0 +1,567 @@
+            raise UserError(_("You cannot modify a posted statement!"))
+
+        if self.parent_id:
+            return
+
+        # clean old lines
+        self.line_ids.unlink()
+
+        # calculate lines
+        lines = self._prepare_lines()
+        move_lines = self._compute_move_lines()
+        self._set_statement_lines(lines, move_lines)
+        move_lines = self._compute_past_invoices_move_lines()
+        self._set_statement_lines(lines, move_lines)
+        self._finalize_lines(lines)
+
+        # create lines
+        self.write(
+            {
+                "line_ids": [(0, 0, line) for line in lines.values()],
+                "date_update": fields.Datetime.now(),
+            }
+        )
+
+    def _compute_past_invoices_move_lines(self):
+        self.ensure_one()
+        moves_to_include = self.unreported_move_ids.filtered(
+            lambda m: m.l10n_nl_vat_statement_include
+        )
+        return moves_to_include.line_ids
+
+    def _compute_move_lines(self):
+        self.ensure_one()
+        domain = self._get_move_lines_domain()
+        return self.env["account.move.line"].search(domain)
+
+    def _set_statement_lines(self, lines, move_lines):
+        self.ensure_one()
+        tags_map = self._get_tags_map()
+        for line in move_lines:
+            for tag in line.tag_ids:
+                tag_map = tags_map.get(tag.id)
+                if tag_map:
+                    code, column = tag_map
+                    if code[0] in ["+", "-"]:
+                        code = code[1:]
+                    lines[code][column] -= line.balance
+
+    def finalize(self):
+        self.ensure_one()
+        self.write({"state": "final"})
+
+    def _check_prev_open_statements(self):
+        prev_open_statements = self.search(
+            [
+                ("company_id", "=", self.company_id.id),
+                ("state", "=", "draft"),
+                ("id", "<", self.id),
+            ],
+            limit=1,
+        )
+        if prev_open_statements:
+            raise UserError(
+                _(
+                    "You cannot post a statement if all the previous "
+                    "statements are not yet posted! "
+                    "Please Post all the other statements first."
+                )
+            )
+
+    def post(self):
+        self.ensure_one()
+        self._check_prev_open_statements()
+
+        self.write({"state": "posted", "date_posted": fields.Datetime.now()})
+        self.unreported_move_ids.filtered(
+            lambda m: m.l10n_nl_vat_statement_include
+        ).write({"l10n_nl_vat_statement_id": self.id})
+        move_lines = self._compute_move_lines()
+        move_lines.move_id.write({"l10n_nl_vat_statement_id": self.id})
+
+    def _get_move_lines_domain(self):
+        company_ids = self._get_company_ids_full_list()
+        domain = [
+            ("company_id", "in", company_ids),
+            ("l10n_nl_vat_statement_id", "=", False),
```

Comment:
@astirpe 
Shouldn't we check on state? Should not include draft and cancelled moves?